### PR TITLE
[SYCLomatic] Refine min/max migration

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 1]
```diff
@@ -813,6 +814,30 @@ void ExprAnalysis::analyzeExpr(const ExplicitCastExpr *Cast) {
   dispatch(Cast->getSubExprAsWritten());
 }
 
+bool isSpecialFunctions(const CallExpr *CE, const std::string &RefString) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]: check name is min/max, and check if this function can be evaluated during migration

File: clang/lib/DPCT/CallExprRewriterMath.cpp [hunk 2]
```diff
@@ -213,29 +204,6 @@ std::string MathFuncNameRewriter::getNewFuncName() {
...
       }
 
-      if (SourceCalleeName == "min" || SourceCalleeName == "max") {
-        LangOptions LO;
-        std::string FT =
-            Call->getType().getCanonicalType().getAsString(PrintingPolicy(LO));
-        for (unsigned i = 0; i < Call->getNumArgs(); i++) {
-          auto Arg = Call->getArg(i);
-          auto ArgExprClass = Arg->getStmtClass();
-          if (isTargetPseudoObjectExpr(Arg)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriterMath.cpp:223]: keep this logic

Comment (zhiweij1) [reply clang/lib/DPCT/CallExprRewriterMath.cpp:223]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/test/dpct/types001.cu [hunk 3]
```diff
@@ -614,12 +614,12 @@ __device__ void foo_3() {
   // CHECK: sycl::range<3> d3 = {3, 2, 1}, *pd3 = &d3;
   dim3 d3 = {1, 2, 3}, *pd3 = &d3;
   int64_t m = 0;
-  // CHECK: m = sycl::min(m, int64_t((*pd3)[2]));
-  // CHECK-NEXT: m = sycl::min(m, int64_t((*pd3)[1]));
-  // CHECK-NEXT: m = sycl::min(m, int64_t((*pd3)[0]));
-  // CHECK-NEXT: m = sycl::min(m, int64_t(d3[2]));
-  // CHECK-NEXT: m = sycl::min(m, int64_t(d3[1]));
-  // CHECK-NEXT: m = sycl::min(m, int64_t(d3[0]));
+  // CHECK: m = dpct::dpct_min(m, int64_t((*pd3)[2]));
```
Comment (zhimingwang36) [inline clang/test/dpct/types001.cu]: this is a regression,  we should use sycl:: one if it works.

File: clang/test/dpct/helper_files_ref/include/math.hpp [hunk 4]
```diff
@@ -164,6 +164,43 @@ inline sycl::half2 isnan(const sycl::half2 h) {
...
+  }                                                                            \
+  inline auto FUNC(const TYPE a, const PROMOTED_TYPE b) {                      \
+    return IMPL_FUNC(static_cast<PROMOTED_TYPE>(a), b);                        \
+  }
+#define DPCT_MATH_MIN_MAX_OVERLOAD_2(FUNC, IMPL_FUNC, TYPE)                    \
+  inline auto FUNC(const TYPE a, const TYPE b) {                               \
+    return IMPL_FUNC(a, b);                                                    \
+  }
+
+DPCT_MATH_MIN_MAX_OVERLOAD_1(dpct_min, sycl::fmin, float, double)
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/math.hpp]: dpct::dpct_min
dpct::math::min()

**Pull request closed without merging**